File: cloudflare-worker/certificate-chain.js
/**
 * Cloudflare Worker for Certificate Chain Analysis
 * Retrieves and analyzes SSL/TLS certificate chains from domains
 * Provides detailed certificate information, chain validation, and security analysis
 */

import { connect } from 'cloudflare:sockets';

// Configuration
const getCertificateConfig = (env) => {
  console.log('getCertificateConfig called with env:', typeof env);
  
  if (!env) {
    throw new Error('Environment object is undefined');
  }
  
  return {
    timeout: parseInt(env.TIMEOUT || '30000'),
    maxRedirects: parseInt(env.MAX_REDIRECTS || '5'),
    userAgent: env.USER_AGENT || 'RussTools Certificate Analyzer/1.0',
    cacheTtl: parseInt(env.CACHE_TTL || '3600')
  };
};

// Allowed origins for CORS
const getAllowedOrigins = (env) => {
  if (!env || !env.ALLOWED_ORIGINS) {
    console.warn('ALLOWED_ORIGINS not configured, using default');
    return ['http://localhost:5173', 'http://localhost:5174', 'https://russ.tools'];
  }
  
  return env.ALLOWED_ORIGINS.split(',').map(origin => origin.trim());
};

// CORS headers
const getCorsHeaders = (origin, allowedOrigins) => {
  const isAllowed = allowedOrigins.includes('*') || allowedOrigins.includes(origin);
  
  return {
    'Access-Control-Allow-Origin': isAllowed ? origin : allowedOrigins[0],
    'Access-Control-Allow-Methods': 'GET, POST, OPTIONS',
    'Access-Control-Allow-Headers': 'Content-Type, Authorization, X-Requested-With',
    'Access-Control-Max-Age': '86400',
  };
};

// Certificate chain retrieval using TLS connection
const retrieveCertificateChain = async (domain, port = 443) => {
  console.log(`Retrieving certificate chain for ${domain}:${port}`);
  
  try {
    // Use Cloudflare Workers TCP sockets API for TLS connection
    const socket = connect({ 
      hostname: domain, 
      port: port 
    }, { 
      secureTransport: "on" 
    });

    // Wait for TLS connection to be established
    const socketInfo = await socket.opened;
    console.log('TLS connection established:', socketInfo);

    // Get certificate information from the TLS socket
    const certificates = await extractRealCertificates(socket, domain, port);
    
    // Close the socket
    await socket.close();
    
    return certificates;
  } catch (error) {
    console.error(`TLS certificate retrieval failed for ${domain}:${port}:`, error);
    throw new Error(`Failed to retrieve certificates from ${domain}:${port} - ${error.message}`);
  }
};

// Extract real certificates from TLS connection
const extractRealCertificates = async (socket, domain, port) => {
  console.log(`Extracting real certificates for ${domain}:${port}`);
  
  try {
    // Get certificate information from the TLS socket
    // Note: The exact API for getting certificates from Cloudflare Workers sockets
    // may not be available yet, so this is a placeholder implementation
    
    throw new Error('Certificate extraction from Cloudflare Workers TCP sockets not yet implemented in the runtime');
    
  } catch (error) {
    console.error(`Real certificate extraction failed:`, error);
    throw new Error(`Failed to extract certificates: ${error.message}`);
  }
};

// NO SSL LABS API - Workers connect() only!

// Parse real certificate from TLS connection
const parseRealCertificate = (cert) => {
  console.log('Parsing real certificate data');
  
  try {
    // Extract certificate information from the actual certificate object from Cloudflare Workers
    return {
      subject: {
        CN: cert.subject?.CN || cert.subject?.commonName || '',
        O: cert.subject?.O || cert.subject?.organizationName || '',
        OU: cert.subject?.OU || cert.subject?.organizationalUnitName || '',
        C: cert.subject?.C || cert.subject?.countryName || '',
        ST: cert.subject?.ST || cert.subject?.stateOrProvinceName || '',
        L: cert.subject?.L || cert.subject?.localityName || '',
        emailAddress: cert.subject?.emailAddress || '',
        raw: cert.subject?.raw || JSON.stringify(cert.subject)
      },
      issuer: {
        CN: cert.issuer?.CN || cert.issuer?.commonName || '',
        O: cert.issuer?.O || cert.issuer?.organizationName || '',
        OU: cert.issuer?.OU || cert.issuer?.organizationalUnitName || '',
        C: cert.issuer?.C || cert.issuer?.countryName || '',
        ST: cert.issuer?.ST || cert.issuer?.stateOrProvinceName || '',
        L: cert.issuer?.L || cert.issuer?.localityName || '',
        emailAddress: cert.issuer?.emailAddress || '',
        raw: cert.issuer?.raw || JSON.stringify(cert.issuer)
      },
      validity: {
        notBefore: new Date(cert.validFrom || cert.valid_from || cert.notBefore),
        notAfter: new Date(cert.validTo || cert.valid_to || cert.notAfter)
      },
      keyInfo: {
        algorithm: cert.signatureAlgorithm || cert.sigAlg || 'Unknown',
        keySize: cert.publicKey?.keySize || cert.keySize || cert.bits || 2048,
        publicKey: cert.publicKey?.raw || new Uint8Array(256)
      },
      extensions: {
        subjectAltName: extractSANs(cert.subjectaltname || cert.altNames || []),
        keyUsage: cert.ext_key_usage || [],
        extendedKeyUsage: cert.ext_key_usage || [],
        basicConstraints: cert.ca ? { isCA: true, pathLenConstraint: null } : null,
        authorityKeyIdentifier: '',
        subjectKeyIdentifier: '',
        certificatePolicies: []
      },
      serialNumber: cert.serialNumber || cert.serial || '',
      version: cert.version || 3
    };
  } catch (error) {
    console.error('Real certificate parsing failed:', error);
    throw new Error(`Failed to parse real certificate: ${error.message}`);
  }
};

// Generate real fingerprints from certificate data
const generateRealFingerprints = async (cert) => {
  console.log('Generating real certificate fingerprints');
  
  try {
    // Use Web Crypto API to generate real fingerprints
    const certBuffer = cert.raw || cert.der || new ArrayBuffer(0);
    
    if (certBuffer.byteLength === 0) {
      throw new Error('No certificate data available for fingerprint generation');
    }
    
    const sha256Hash = await crypto.subtle.digest('SHA-256', certBuffer);
    const sha1Hash = await crypto.subtle.digest('SHA-1', certBuffer);
    
    return {
      sha256: Array.from(new Uint8Array(sha256Hash))
        .map(b => b.toString(16).padStart(2, '0'))
        .join(':'),
      sha1: Array.from(new Uint8Array(sha1Hash))
        .map(b => b.toString(16).padStart(2, '0'))
        .join(':')
    };
  } catch (error) {
    console.error('Real fingerprint generation failed:', error);
    throw new Error(`Failed to generate real fingerprints: ${error.message}`);
  }
};

// Extract subject/issuer information from certificate field
const extractSubjectInfo = (subjectField) => {
  if (typeof subjectField === 'string') {
    // Parse DN string format
    const parts = subjectField.split(',').map(p => p.trim());
    const subject = { raw: subjectField };
    
    parts.forEach(part => {
      const [key, value] = part.split('=').map(p => p.trim());
      if (key && value) {
        subject[key] = value;
      }
    });
    
    return {
      CN: subject.CN || '',
      O: subject.O || '',
      OU: subject.OU || '',
      C: subject.C || '',
      ST: subject.ST || subject.S || '',
      L: subject.L || '',
      emailAddress: subject.emailAddress || subject.E || '',
      raw: subjectField
    };
  }
  
  // Handle object format
  return {
    CN: subjectField?.commonName || subjectField?.CN || '',
    O: subjectField?.organizationName || subjectField?.O || '',
    OU: subjectField?.organizationalUnitName || subjectField?.OU || '',
    C: subjectField?.countryName || subjectField?.C || '',
    ST: subjectField?.stateOrProvinceName || subjectField?.ST || '',
    L: subjectField?.localityName || subjectField?.L || '',
    emailAddress: subjectField?.emailAddress || '',
    raw: subjectField?.raw || JSON.stringify(subjectField)
  };
};

// Extract Subject Alternative Names
const extractSANs = (sanData) => {
  if (Array.isArray(sanData)) {
    return sanData.map(san => {
      if (typeof san === 'string') {
        return { type: 'DNS', value: san };
      }
      return { type: san.type || 'DNS', value: san.value || san.name || san };
    });
  }
  return [];
};

// NO SSL LABS HELPERS - Workers connect() only!

// REMOVED - No mock certificate parsing

// Determine certificate type in chain
const determineCertificateType = (cert, allCerts, index) => {
  console.log(`Determining certificate type for certificate ${index}`);
  
  const isRoot = cert.subject.raw === cert.issuer.raw;
  const isLast = index === allCerts.length - 1;
  const isFirst = index === 0;
  
  if (isRoot) return 'root';
  if (isFirst) return 'leaf';
  if (!isFirst && !isLast) return 'intermediate';
  
  return 'intermediate';
};

// REMOVED - No mock fingerprint generation

// Analyze certificate for security issues
const analyzeCertificateIssues = (cert) => {
  console.log('Analyzing certificate for security issues');
  
  const issues = [];
  const now = new Date();
  
  // Check expiration
  const daysUntilExpiry = Math.ceil((cert.validity.notAfter - now) / (1000 * 60 * 60 * 24));
  if (daysUntilExpiry <= 0) {
    issues.push({
      severity: 'critical',
      type: 'expired',
      description: `Certificate expired ${Math.abs(daysUntilExpiry)} days ago`
    });
  } else if (daysUntilExpiry <= 30) {
    issues.push({
      severity: 'warning',
      type: 'expiring_soon',
      description: `Certificate expires in ${daysUntilExpiry} days`
    });
  }
  
  // Check key strength
  if (cert.keyInfo.keySize < 2048) {
    issues.push({
      severity: 'critical',
      type: 'weak_key',
      description: `Weak key size: ${cert.keyInfo.keySize} bits`
    });
  }
  
  // Check signature algorithm
  if (cert.keyInfo.algorithm.includes('SHA1')) {
    issues.push({
      severity: 'warning',
      type: 'weak_signature',
      description: 'Uses weak SHA-1 signature algorithm'
    });
  }
  
  if (cert.keyInfo.algorithm.includes('MD5')) {
    issues.push({
      severity: 'critical',
      type: 'insecure_signature',
      description: 'Uses insecure MD5 signature algorithm'
    });
  }
  
  return issues;
};

// NO MOCK DATA - Real certificates only!

// Validate certificate chain
const validateCertificateChain = (certificates) => {
  console.log('Validating certificate chain');
  
  const validation = {
    isValid: true,
    trustPath: certificates,
    issues: [],
    metadata: {
      chainLength: certificates.length,
      validFrom: null,
      validTo: null,
      hasRoot: false,
      hasIntermediate: false,
      hasLeaf: false
    }
  };
  
  // Update metadata
  validation.metadata.hasRoot = certificates.some(cert => cert.type === 'root');
  validation.metadata.hasIntermediate = certificates.some(cert => cert.type === 'intermediate');
  validation.metadata.hasLeaf = certificates.some(cert => cert.type === 'leaf');
  
  const leafCert = certificates.find(cert => cert.type === 'leaf');
  if (leafCert) {
    validation.metadata.validFrom = leafCert.details.validity.notBefore;
    validation.metadata.validTo = leafCert.details.validity.notAfter;
  }
  
  // Basic chain validation
  if (!validation.metadata.hasLeaf) {
    validation.issues.push({
      severity: 'error',
      type: 'missing_leaf',
      description: 'No leaf certificate found in chain'
    });
    validation.isValid = false;
  }
  
  if (!validation.metadata.hasRoot) {
    validation.issues.push({
      severity: 'warning',
      type: 'incomplete_chain',
      description: 'No root certificate found in chain'
    });
  }
  
  return validation;
};

// Main request handler
const worker = {
  async fetch(request, env) {
    console.log('Certificate Chain Worker request received:', request.method, request.url);
    
    try {
      const config = getCertificateConfig(env);
      const allowedOrigins = getAllowedOrigins(env);
      const origin = request.headers.get('origin');
      const corsHeaders = getCorsHeaders(origin, allowedOrigins);
      
      // Handle CORS preflight
      if (request.method === 'OPTIONS') {
        return new Response(null, {
          status: 200,
          headers: corsHeaders
        });
      }
      
      // Handle health check
      if (request.url.endsWith('/health')) {
        return new Response(JSON.stringify({
          status: 'healthy',
          timestamp: new Date().toISOString(),
          version: '1.0.0',
          service: 'certificate-chain-analyzer'
        }), {
          status: 200,
          headers: {
            'Content-Type': 'application/json',
            ...corsHeaders
          }
        });
      }
      
      // Handle certificate analysis
      if (request.method === 'POST' && request.url.endsWith('/analyze')) {
        const requestData = await request.json();
        const { domain, port = 443 } = requestData;
        
        if (!domain) {
          return new Response(JSON.stringify({
            success: false,
            error: 'Domain parameter is required'
          }), {
            status: 400,
            headers: {
              'Content-Type': 'application/json',
              ...corsHeaders
            }
          });
        }
        
        console.log(`Analyzing certificate chain for ${domain}:${port}`);
        
        try {
          const certificates = await retrieveCertificateChain(domain, port);
          const chainValidation = validateCertificateChain(certificates);
          
          const response = {
            success: true,
            domain,
            port,
            timestamp: new Date().toISOString(),
            certificates,
            chain: chainValidation,
            metadata: {
              totalCertificates: certificates.length,
              analysisType: 'cloudflare-worker',
              source: `${domain}:${port}`,
              workerVersion: '1.0.0'
            }
          };
          
          return new Response(JSON.stringify(response), {
            status: 200,
            headers: {
              'Content-Type': 'application/json',
              'Cache-Control': `public, max-age=${config.cacheTtl}`,
              ...corsHeaders
            }
          });
          
        } catch (error) {
          console.error('Certificate analysis failed:', error);
          
          return new Response(JSON.stringify({
            success: false,
            error: error.message,
            domain,
            port,
            timestamp: new Date().toISOString()
          }), {
            status: 500,
            headers: {
              'Content-Type': 'application/json',
              ...corsHeaders
            }
          });
        }
      }
      
      // Invalid endpoint
      return new Response(JSON.stringify({
        success: false,
        error: 'Not found',
        availableEndpoints: ['/health', '/analyze']
      }), {
        status: 404,
        headers: {
          'Content-Type': 'application/json',
          ...corsHeaders
        }
      });
      
    } catch (error) {
      console.error('Worker error:', error);
      
      return new Response(JSON.stringify({
        success: false,
        error: 'Internal server error',
        timestamp: new Date().toISOString()
      }), {
        status: 500,
        headers: {
          'Content-Type': 'application/json',
          'Access-Control-Allow-Origin': '*'
        }
      });
    }
  }
};

export default worker;

File: cloudflare-worker/certificate-chain.py
"""
Cloudflare Python Worker for Certificate Chain Analysis
Retrieves and analyzes SSL/TLS certificate chains from domains
Uses Python's ssl and socket modules for real certificate extraction
"""

import ssl
import socket
import json
import hashlib
from datetime import datetime
from typing import Dict, List, Any, Optional

def application(environ, start_response):
    """WSGI application for certificate chain analysis"""
    
    # Handle CORS
    cors_headers = [
        ('Access-Control-Allow-Origin', '*'),
        ('Access-Control-Allow-Methods', 'GET, POST, OPTIONS'),
        ('Access-Control-Allow-Headers', 'Content-Type, Authorization, X-Requested-With'),
        ('Access-Control-Max-Age', '86400'),
        ('Content-Type', 'application/json')
    ]
    
    method = environ.get('REQUEST_METHOD', 'GET')
    path = environ.get('PATH_INFO', '/')
    
    try:
        # Handle CORS preflight
        if method == 'OPTIONS':
            start_response('200 OK', cors_headers)
            return [b'']
        
        # Health check endpoint
        if path.endswith('/health'):
            response = {
                'status': 'healthy',
                'timestamp': datetime.utcnow().isoformat() + 'Z',
                'version': '1.0.0',
                'service': 'certificate-chain-analyzer-python',
                'language': 'python'
            }
            start_response('200 OK', cors_headers)
            return [json.dumps(response).encode('utf-8')]
        
        # Certificate analysis endpoint
        if method == 'POST' and path.endswith('/analyze'):
            # Read request body
            content_length = int(environ.get('CONTENT_LENGTH', '0'))
            if content_length == 0:
                raise ValueError('Request body is required')
            
            body = environ['wsgi.input'].read(content_length)
            request_data = json.loads(body.decode('utf-8'))
            
            domain = request_data.get('domain')
            port = request_data.get('port', 443)
            
            if not domain:
                raise ValueError('Domain parameter is required')
            
            print(f"Analyzing certificate chain for {domain}:{port}")
            
            # Retrieve and analyze certificate chain
            certificates = retrieve_certificate_chain(domain, port)
            chain_validation = validate_certificate_chain(certificates)
            
            response = {
                'success': True,
                'domain': domain,
                'port': port,
                'timestamp': datetime.utcnow().isoformat() + 'Z',
                'certificates': certificates,
                'chain': chain_validation,
                'metadata': {
                    'totalCertificates': len(certificates),
                    'analysisType': 'python-worker',
                    'source': f'{domain}:{port}',
                    'workerVersion': '1.0.0'
                }
            }
            
            start_response('200 OK', cors_headers)
            return [json.dumps(response).encode('utf-8')]
        
        # Not found
        response = {
            'success': False,
            'error': 'Not found',
            'availableEndpoints': ['/health', '/analyze']
        }
        start_response('404 Not Found', cors_headers)
        return [json.dumps(response).encode('utf-8')]
        
    except Exception as error:
        print(f"Worker error: {error}")
        
        response = {
            'success': False,
            'error': str(error),
            'timestamp': datetime.utcnow().isoformat() + 'Z'
        }
        start_response('500 Internal Server Error', cors_headers)
        return [json.dumps(response).encode('utf-8')]


def retrieve_certificate_chain(domain: str, port: int = 443) -> List[Dict[str, Any]]:
    """Retrieve real certificate chain using Python's ssl module"""
    print(f"Retrieving certificate chain for {domain}:{port}")
    
    try:
        # Create SSL context
        context = ssl.create_default_context()
        
        # Create socket connection
        with socket.create_connection((domain, port), timeout=30) as sock:
            # Wrap socket with SSL
            with context.wrap_socket(sock, server_hostname=domain) as ssock:
                # Get peer certificate chain
                cert_der = ssock.getpeercert(binary_form=True)
                cert_pem = ssl.DER_cert_to_PEM_cert(cert_der)
                
                # Get certificate details
                cert_details = ssock.getpeercert()
                
                # Get certificate chain if available
                cert_chain = ssock.getpeercert_chain()
                
                certificates = []
                
                if cert_chain:
                    # Process full certificate chain
                    for i, cert in enumerate(cert_chain):
                        cert_der_bytes = cert.public_bytes(encoding=ssl.Encoding.DER)
                        cert_details_dict = cert.to_dict() if hasattr(cert, 'to_dict') else {}
                        
                        parsed_cert = parse_certificate(cert_der_bytes, cert_details_dict, domain)
                        cert_type = determine_certificate_type(parsed_cert, cert_chain, i)
                        
                        certificates.append({
                            'id': f'cert-{i}',
                            'type': cert_type,
                            'raw': cert_der_bytes.hex(),
                            'details': parsed_cert,
                            'fingerprints': generate_fingerprints(cert_der_bytes),
                            'issues': analyze_certificate_issues(parsed_cert)
                        })
                else:
                    # Single certificate only
                    parsed_cert = parse_certificate(cert_der, cert_details, domain)
                    
                    certificates.append({
                        'id': 'cert-0',
                        'type': 'leaf',
                        'raw': cert_der.hex(),
                        'details': parsed_cert,
                        'fingerprints': generate_fingerprints(cert_der),
                        'issues': analyze_certificate_issues(parsed_cert)
                    })
                
                return certificates
                
    except Exception as error:
        print(f"Certificate retrieval failed for {domain}:{port}: {error}")
        raise Exception(f"Failed to retrieve certificates from {domain}:{port} - {error}")


def parse_certificate(cert_der: bytes, cert_details: Dict, domain: str) -> Dict[str, Any]:
    """Parse certificate details from DER bytes and certificate info"""
    
    try:
        # Extract subject information
        subject = cert_details.get('subject', ())
        subject_dict = dict(x[0] for x in subject)
        
        # Extract issuer information  
        issuer = cert_details.get('issuer', ())
        issuer_dict = dict(x[0] for x in issuer)
        
        # Extract validity dates
        not_before = cert_details.get('notBefore', '')
        not_after = cert_details.get('notAfter', '')
        
        # Parse dates
        not_before_dt = datetime.strptime(not_before, '%b %d %H:%M:%S %Y %Z') if not_before else datetime.utcnow()
        not_after_dt = datetime.strptime(not_after, '%b %d %H:%M:%S %Y %Z') if not_after else datetime.utcnow()
        
        # Extract subject alternative names
        san_list = []
        for name, value in cert_details.get('subjectAltName', []):
            san_list.append({'type': name, 'value': value})
        
        return {
            'subject': {
                'CN': subject_dict.get('commonName', ''),
                'O': subject_dict.get('organizationName', ''),
                'OU': subject_dict.get('organizationalUnitName', ''),
                'C': subject_dict.get('countryName', ''),
                'ST': subject_dict.get('stateOrProvinceName', ''),
                'L': subject_dict.get('localityName', ''),
                'emailAddress': subject_dict.get('emailAddress', ''),
                'raw': ', '.join(f'{k}={v}' for k, v in subject_dict.items())
            },
            'issuer': {
                'CN': issuer_dict.get('commonName', ''),
                'O': issuer_dict.get('organizationName', ''),
                'OU': issuer_dict.get('organizationalUnitName', ''),
                'C': issuer_dict.get('countryName', ''),
                'ST': issuer_dict.get('stateOrProvinceName', ''),
                'L': issuer_dict.get('localityName', ''),
                'emailAddress': issuer_dict.get('emailAddress', ''),
                'raw': ', '.join(f'{k}={v}' for k, v in issuer_dict.items())
            },
            'validity': {
                'notBefore': not_before_dt.isoformat() + 'Z',
                'notAfter': not_after_dt.isoformat() + 'Z'
            },
            'keyInfo': {
                'algorithm': 'Unknown',  # Would need pyOpenSSL for detailed info
                'keySize': 2048,  # Default assumption
                'publicKey': cert_der[:256].hex()  # First 256 bytes as placeholder
            },
            'extensions': {
                'subjectAltName': san_list,
                'keyUsage': [],
                'extendedKeyUsage': [],
                'basicConstraints': None,
                'authorityKeyIdentifier': '',
                'subjectKeyIdentifier': '',
                'certificatePolicies': []
            },
            'serialNumber': str(cert_details.get('serialNumber', '')),
            'version': cert_details.get('version', 3)
        }
        
    except Exception as error:
        print(f"Certificate parsing failed: {error}")
        raise Exception(f"Failed to parse certificate: {error}")


def determine_certificate_type(cert: Dict, cert_chain: List, index: int) -> str:
    """Determine if certificate is leaf, intermediate, or root"""
    
    subject_cn = cert['subject']['CN']
    issuer_cn = cert['issuer']['CN']
    
    # Root certificate: subject == issuer
    if subject_cn == issuer_cn:
        return 'root'
    
    # Leaf certificate: first in chain
    if index == 0:
        return 'leaf'
    
    # Everything else is intermediate
    return 'intermediate'


def generate_fingerprints(cert_der: bytes) -> Dict[str, str]:
    """Generate SHA256 and SHA1 fingerprints from certificate DER bytes"""
    
    try:
        sha256_hash = hashlib.sha256(cert_der).hexdigest()
        sha1_hash = hashlib.sha1(cert_der).hexdigest()
        
        # Format as colon-separated hex
        sha256_formatted = ':'.join(sha256_hash[i:i+2] for i in range(0, len(sha256_hash), 2))
        sha1_formatted = ':'.join(sha1_hash[i:i+2] for i in range(0, len(sha1_hash), 2))
        
        return {
            'sha256': sha256_formatted,
            'sha1': sha1_formatted
        }
        
    except Exception as error:
        print(f"Fingerprint generation failed: {error}")
        return {'sha256': '', 'sha1': ''}


def analyze_certificate_issues(cert: Dict) -> List[Dict[str, str]]:
    """Analyze certificate for security issues"""
    
    issues = []
    
    try:
        # Check expiration
        not_after_str = cert['validity']['notAfter']
        not_after = datetime.fromisoformat(not_after_str.replace('Z', '+00:00'))
        now = datetime.utcnow().replace(tzinfo=not_after.tzinfo)
        
        days_until_expiry = (not_after - now).days
        
        if days_until_expiry <= 0:
            issues.append({
                'severity': 'critical',
                'type': 'expired',
                'description': f'Certificate expired {abs(days_until_expiry)} days ago'
            })
        elif days_until_expiry <= 30:
            issues.append({
                'severity': 'warning',
                'type': 'expiring_soon',
                'description': f'Certificate expires in {days_until_expiry} days'
            })
        
        # Check key strength
        key_size = cert['keyInfo']['keySize']
        if key_size < 2048:
            issues.append({
                'severity': 'critical',
                'type': 'weak_key',
                'description': f'Weak key size: {key_size} bits'
            })
        
        # Check algorithm
        algorithm = cert['keyInfo']['algorithm']
        if 'SHA1' in algorithm:
            issues.append({
                'severity': 'warning',
                'type': 'weak_signature',
                'description': 'Uses weak SHA-1 signature algorithm'
            })
        elif 'MD5' in algorithm:
            issues.append({
                'severity': 'critical',
                'type': 'insecure_signature',
                'description': 'Uses insecure MD5 signature algorithm'
            })
            
    except Exception as error:
        print(f"Certificate issue analysis failed: {error}")
    
    return issues


def validate_certificate_chain(certificates: List[Dict]) -> Dict[str, Any]:
    """Validate certificate chain structure and trust path"""
    
    validation = {
        'isValid': True,
        'trustPath': certificates,
        'issues': [],
        'metadata': {
            'chainLength': len(certificates),
            'validFrom': None,
            'validTo': None,
            'hasRoot': False,
            'hasIntermediate': False,
            'hasLeaf': False
        }
    }
    
    # Check certificate types
    for cert in certificates:
        cert_type = cert['type']
        if cert_type == 'root':
            validation['metadata']['hasRoot'] = True
        elif cert_type == 'intermediate':
            validation['metadata']['hasIntermediate'] = True
        elif cert_type == 'leaf':
            validation['metadata']['hasLeaf'] = True
            # Set validity period from leaf certificate
            validation['metadata']['validFrom'] = cert['details']['validity']['notBefore']
            validation['metadata']['validTo'] = cert['details']['validity']['notAfter']
    
    # Validate chain structure
    if not validation['metadata']['hasLeaf']:
        validation['issues'].append({
            'severity': 'error',
            'type': 'missing_leaf',
            'description': 'No leaf certificate found in chain'
        })
        validation['isValid'] = False
    
    if not validation['metadata']['hasRoot']:
        validation['issues'].append({
            'severity': 'warning',
            'type': 'incomplete_chain',
            'description': 'No root certificate found in chain'
        })
    
    return validation